### ISSUE TITLE
Paypal completing payment but returning to checkout not order conf, order not complete in OFN

### ISSUE THREAD

**Issue by kirstenalarsen** [openfoodfoundation/openfoodnetwork#1837]
User reported: My store was open Thursday - Sunday yesterday and I took like 8 orders however there is no evidence or confirmation in OFN. Only PayPal receipts with orders in there. 

The orders are in OFN but are not completed, still in State 'checkout'

![image](https://user-images.githubusercontent.com/2911176/30531167-ada96d4e-9c8f-11e7-8d6c-96b9ff708ecf.png)

I have confirmed this problem when I shop through their shop (the mandarin bend) but that our test paypal account on prod is fine

**Comment by kirstenalarsen** [comment]
Have now confirmed that when I put a payment fee on it causes this problem, so very likely connected to #1074 

However, I think this behaviour in response to that problem is new? @oeoeaio @mkllnk @sstead

**Comment by oeoeaio** [comment]
Probably not directly related, but great information to have!

Additional information (from server logs):
````
Redirected by /home/ubuntu/.gem/ruby/2.1.0/bundler/gems/better_spree_paypal_express-8d95f4544c68/app/controllers/spree/paypal_controller.rb:62:in `confirm'
````

This is the method:
````ruby
# app/controllers/spree/paypal_controller.rb
def confirm
  order = current_order || raise(ActiveRecord::RecordNotFound)
  order.payments.create!({
    :source => Spree::PaypalExpressCheckout.create({
      :token => params[:token],
      :payer_id => params[:PayerID]
    }, :without_protection => true),
    :amount => order.total,
    :payment_method => payment_method
  }, :without_protection => true)
  order.next
  if order.complete?
    flash.notice = Spree.t(:order_processed_successfully)
    flash[:commerce_tracking] = "nothing special"
    session[:order_id] = nil
    redirect_to completion_route(order)
  else
    redirect_to checkout_state_path(order.state) # This is line 62
  end
end
````

So now all we need to do is to work out is why adding a payment fee causes the order to not be complete at this point...

**Comment by oeoeaio** [comment]
Ok, so maybe it is more related to #1074 than I thought: `order.next` is failing to transition here because of [this callback](https://github.com/coopdevs/spree/blob/spree-upgrade-step-6/core/app/models/spree/order/checkout.rb#L66) for the state machine on `order`

More specifically: there are two payments on the order at the time `order.next` is called in the `confirm` action above, one is created earlier in the checkout process and one is created in the `confirm` action itself. The one that was created earlier has no source and so it fails [this check](https://github.com/coopdevs/spree/blob/spree-upgrade-step-6/core/app/models/spree/payment/processing.rb#L6), resulting in a `Spree::Core::GatewayError` being raise, and the transition of the order to `complete` to fail.

**Comment by oeoeaio** [comment]
Ok, two more pieces of information:

1. The first payment is created in the CheckoutController by the params submitted in `payments_attributes`. For PayPal payments the only attributes submitted are `amount` and `payment_method_id`. Only `payment_method_id` is actually used, because the `PayPalController#express` action from `better_spree_paypal_express` ignores `amount` and recalculates the total to be charged from scratch. We could prevent this payment from being created by simply stripping out the `payments_attributes` from the params before @order.updates_attributes is called. The only problem with this is that is leaves us without anything to attach a transaction fee to, and the second payment is not created until after the payment amount has been confirmed with PayPal...

2. For payments that don't have a transaction fee, there are still two payments on the order, but for some reason the order that they are returned in is reversed, so that the newer payment appears first and the older payment appears second. This means that the newer payment is processed first by order.process_payments! which has a `break` clause `if payment_total >= total` which is true after the newer payment is processed, so no attempt to process the older (broken payment without a source) is made.

**Comment by oeoeaio** [comment]
For posterity, here is my full understanding of this issue as it currently stands.

Starting at the start: in 2014, we rebuilt the checkout flow, so that it had a better UX. This involved submitting all attributes of the order in one step, so we now submit `payments_attributes` with every request made to `CheckoutController#update`. Generally speaking these attributes contain all of the information required for the server to process the payment (card number and expiry for PinPayments, a token for Stripe). The `payment_method_id` and an `amount` are always submitted.

Then we integrated the PayPal gateway (using `spree_paypal_express`. With PalPal payments, the `payment_method_id` and `amount` are the only things we submit (the amount is ignored). The customer is redirected to PayPal (via `spree_paypal_express`), which issues a token that we can use to process the payment. That worked fine, and the fact that we were submitting a `payments_attributes` hash that is basically empty never caused a major problem in an of itself. As far as I know, 'orphaned' payments were created in `CheckoutController#update` from the very beginning, but `spree_paypal_express` always ignored these and created a new payment, without this causing an issue.

At some point we added the ability to apply transaction fees. I didn't do this work, but I suspect that the way that this works for PayPal payments wasn't really 'designed' in a deliberate way, it just 'happened' to work. The way it 'works' is that the first payment (created from `payments_attributes` params in `CheckoutController#update`) creates a transaction fee on the order. Then `spree_paypal_express` includes this fee in the order total that is submitted to PayPal for confirmation by the user (see `PayPalController#express` for this calculation logic). When the user is returned to the OFN a new payment is created (by `PayPalController#confirm`) and the user is redirected to the final order confirmation page. This results in two payments, one that is 'orphaned ' and one that is processed, both payments create transaction fees on the order. The 'orphaned' payment is for the correct amount, while the processed payment total includes both it's own transaction fees and the 'orphaned' transaction fees. The customer is only actually charged for the first transaction fee (because the second one was only created after the total was confirmed). The order total includes both transaction fees, but because the processed payment total (stored locally) is for the same amount, the payment total matches the order total. So the data is a complete mess. This is basically what issue #1074 is about.

I'm not exactly sure what happened in the spree upgrade but for some reason, this finely balanced setup that sort-of-worked decided to stopped working. The crux of the issue is this: `order.process_payments!` is called by the `before_transition :to => :complete` callback in `Spree::Order::Checkout`, which is in turn initiated `order.next` immediately after the payment is created in `PayPalController#confirm`. `order.process_payments!` used to run fine, and never attempted to process the 'orphaned' payment. With the latest Spree upgrade however, an attempt it made to process it, which causes an error (since there is no payment information on it to process, specifically: it has no `source`), which causes the transition of the order to the `complete` state to fail, and the customer is then redirected to the checkout.

There is a `break` clause in `order.process_payments!` that stops processing as soon as `payment_total >= total` is `true`. I initially thought that the issue was simply caused by some weirdness in the order that the payments were being loaded for processing, but I think that was a red herring. I suspect the real issue is related to the calculation of the totals. As far as I can gather, what is happening now is that the 'real' payment created by `spree_paypal_express` in `PayPalController#confirm` is for the **correct** amount (ie. doesn't include the transaction fee twice). However, the order total is still for the wrong amount, so once it has been processed, the `break` clause is not satisfied and an attempt is made to process the second 'orphaned' payment (which fails).

My PR #1842 fixes the issue by making sure that both the only one payment and transaction fee exist on the order at the time processing occurs. This also has the benefit of fixing the worst of the issues reported in #1074. It's not particularly elegant: we create a payment and transaction fee that are essentially placeholders to enable correct calculation of the total for submission to PayPal, which are later replaced by the rightful payment and transaction fee. So the process is still a bit of a mess, but it's working much better than it ever was before, and there just aren't the resources or impetus to come up with a better fix at this stage.
